posts/20-candid-for-engineers.html.pm

```diff
@@ -0,0 +1,908 @@
+  add_user : (User) -> (nat);
+  get_user : (nat) -> (User) query;
+ }
+}
+
+◊p{
+  The answer depends on the record type variance if the field is not ◊code{opt}.
+}
+
+◊p{
+  You can remove the field if the type appears only in method arguments but prefer marking it as ◊code{reserved} instead.
+}
+
+◊source-code["good"]{
+ service UserService : {
+-  add_user : (record { name : text;  age : nat }) -> (nat);
++  add_user : (record { name : text             }) -> (nat);
+ }
+}
+
+◊source-code["good"]{
+ service UserService : {
+-  add_user : (record { name : text; age : nat      }) -> (nat);
++  add_user : (record { name : text; ◊b{age : reserved} }) -> (nat);
+ }
+}
+
+◊p{
+   You should preserve the field if the type appears in a method return type.
+}
+
+◊source-code["bad"]{
+◊em{// BAD: the User type appears as an argument ◊b{and} a result.}
+ type User = record {
+   name : text;
+-  age : nat;
+};
+
+ service UserService : {
+  add_user : (User) -> (nat);
+  get_user : (nat) -> (User) query;
+ }
+}
+
+◊subsection-title["faq-remove-field"]{Can I add a record field?}
+
+◊p{
+  Adding an ◊code{opt} field is always safe.
+}
+
+◊source-code["good"]{
+ type User = record {
+   name : text;
++  age : opt nat;
+};
+
+ service UserService : {
+  add_user : (User) -> (nat);
+  get_user : (nat) -> (User) query;
+ }
+}
+
+◊p{
+  For non-◊code{opt} fields, the answer depends on the type variance.
+}
+
+◊p{
+  You can safely add a non-optional field if the record appears only in method return types.
+}
+
+◊source-code["good"]{
+ ◊em{// BAD: breaks the client code}
+ service UserService : {
+-  get_user : (nat) -> (record { name : text            }) query;
++  get_user : (nat) -> (record { name : text; ◊b{age : nat} }) query;
+ }
+}
+
+◊p{
+  Adding a non-optional field breaks backward compatibility if the record appears in a method argument.
+}
+
+◊source-code["bad"]{
+ service UserService : {
+-  add_user : (record { name : text            }) -> (nat);
++  add_user : (record { name : text; ◊b{age : nat} }) -> (nat);
+ }
+}
+
+
+◊subsection-title["faq-remove-alternative"]{Can I remove a variant alternative?}
+
+◊p{
+  It depends on the type variance.
+}
+
+◊p{
+  You can remove alternatives if the variant appears only in method results.
+}
+
+◊source-code["good"]{
+ service CoffeeShop : {
+-  order_size : (nat) -> (variant { tiny; small; medium; large }) query;
++  order_size : (nat) -> (variant {       small; medium; large }) query;
+ }
+}
+
+◊source-code["bad"]{
+ ◊em{// BAD: this change might break clients.}
+ service CoffeeShop : {
+-  order_coffee : (record { size : variant { tiny; small; medium; large } }) -> (nat);
++  order_coffee : (record { size : variant {       small; medium; large } }) -> (nat);
+ }
+}
+
+
+◊subsection-title["faq-remove-alternative"]{Can I add a variant alternative?}
+
+◊p{
+  It depends on the type variance.
+}
+
+◊p{
+  If the variant appears only in method arguments, you can safely add new alternatives.
```

Comment:
We may also mention the `opt variant` case, where you can safely add and remove variant tags.